Source/BugsnagSessionTracker.m

```diff
@@ -41,92 +50,61 @@ - (instancetype)initWithConfig:(BugsnagConfiguration *)config
     return self;
 }
 
-- (void)startNewSession:(NSDate *)date
-               withUser:(BugsnagUser *)user
-           autoCaptured:(BOOL)autoCaptured {
-    if (self.config.sessionURL == nil) {
-        bsg_log_err(@"The session tracking endpoint has not been set. Session tracking is disabled");
-        return;
-    }
+#pragma mark - Creating and sending a new session
 
-    _currentSession = [[BugsnagSession alloc] initWithId:[[NSUUID UUID] UUIDString]
-                                                startDate:date
-                                                     user:user
-                                             autoCaptured:autoCaptured];
+- (void)startNewSession {
+    [self startNewSessionWithAutoCaptureValue:NO];
+}
 
-    if ((self.config.shouldAutoCaptureSessions || !autoCaptured) && [self.config shouldSendReports]) {
-        [self trackSession];
+- (void)startNewSessionIfAutoCaptureEnabled {
+    if (self.config.shouldAutoCaptureSessions  && [self.config shouldSendReports]) {
+        [self startNewSessionWithAutoCaptureValue:YES];
     }
-    _isInForeground = YES;
 }
 
-- (void)trackSession {
+- (void)startNewSessionWithAutoCaptureValue:(BOOL)isAutoCaptured {
+    if (self.config.sessionURL == nil) {
+        bsg_log_err(@"The session tracking endpoint has not been set. Session tracking is disabled");
+        return;
+    }
+    self.currentSession = [[BugsnagSession alloc] initWithId:[[NSUUID UUID] UUIDString]
+                                                   startDate:[NSDate date]
+                                                        user:self.config.currentUser
+                                                autoCaptured:isAutoCaptured];
+
     [self.sessionStore write:self.currentSession];
-    self.trackedFirstSession = YES;
 
     if (self.callback) {
         self.callback(self.currentSession);
     }
+    [self.apiClient deliverSessionsInStore:self.sessionStore];
 }
 
-- (void)onAutoCaptureEnabled {
-    if (!self.trackedFirstSession) {
-        if (self.currentSession == nil) { // unlikely case, will be initialised later
-            return;
-        }
-        [self trackSession];
-    }
-}
+#pragma mark - Handling events
 
-- (void)suspendCurrentSession:(NSDate *)date {
-    _isInForeground = NO;
+- (void)handleAppBackgroundEvent {
+    self.backgroundStartTime = [NSDate date];
 }
 
-- (void)incrementHandledError {
-    @synchronized (self.currentSession) {
-        self.currentSession.handledCount++;
-        if (self.callback && (self.config.shouldAutoCaptureSessions || !self.currentSession.autoCaptured)) {
-            self.callback(self.currentSession);
-        }
+- (void)handleAppForegroundEvent {
+    if (self.backgroundStartTime
+        && [[NSDate date] timeIntervalSinceDate:self.backgroundStartTime] >= BSGNewSessionBackgroundDuration) {
+        [self startNewSessionIfAutoCaptureEnabled];
     }
+    self.backgroundStartTime = nil;
```

Comment:
If this can be nil and we're expecting sessions to be started from multiple threads, would it make sense to make this property atomic/make any checks compound operations?